src/solvers/dg_simplices/dg.jl

```diff
@@ -0,0 +1,301 @@
+
+@inline eachdim(mesh) = Base.OneTo(ndims(mesh)) 
+
+# iteration over all elements in a mesh
+@inline ndofs(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = dg.basis.Np * mesh.md.num_elements
+@inline eachelement(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(mesh.md.num_elements)
+
+# iteration over quantities in a single element
+@inline each_face_node(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(dg.basis.Nfq)
+@inline each_quad_node(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(dg.basis.Nq)
+
+# iteration over quantities over the entire mesh (dofs, quad nodes, face nodes). 
+@inline each_dof_global(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(ndofs(mesh, dg, cache))
+@inline each_quad_node_global(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(dg.basis.Nq * mesh.md.num_elements)
+@inline each_face_node_global(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(dg.basis.Nfq * mesh.md.num_elements)
+
+# interface with semidiscretization_hyperbolic
+wrap_array(u_ode::StructArray, mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache) = u_ode
+function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, mesh::AbstractMeshData, 
+                                    dg::DG{<:RefElemData}, cache) where {Keys,ValueTypes<:NTuple{N,Any}} where {N}
+  return boundary_conditions
+end
+
+function allocate_coefficients(mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache)
+  md = mesh.md
+  nvars = nvariables(equations) 
+  return StructArray{SVector{nvars, real(dg)}}(ntuple(_->similar(md.x),nvars))
+end
+
+function compute_coefficients!(u::StructArray, initial_condition, t,
+                        mesh::AbstractMeshData{Dim}, equations, dg::DG{<:RefElemData{Dim}}, cache) where {Dim}
+  md = mesh.md
+  rd = dg.basis
+  @unpack u_values = cache
+
+  @threaded for i in each_quad_node_global(mesh, dg, cache)    
+    u_values[i] = initial_condition(getindex.(md.xyzq, i), t, equations) 
+  end
+
+  # compute L2 projection
+  StructArrays.foreachfield(mul_by!(rd.Pq), u, u_values)
+end
+
+# interpolates from solution coefficients to face quadrature points
+function prolong2interfaces!(cache, u, mesh::AbstractMeshData, equations, 
+                             surface_integral, dg::DG{<:RefElemData})
+  rd = dg.basis    
+  @unpack u_face_values = cache
+  StructArrays.foreachfield(mul_by!(rd.Vf), u_face_values, u)
+end
+
+function create_cache(mesh::VertexMappedMesh, equations, dg::DG, 
+                      RealT, uEltype) where {DG <: DGWeakForm{Dim}} where {Dim}
+
+  rd = dg.basis
+  md = mesh.md
+
+  # volume quadrature weights, volume interpolation matrix
+  @unpack wq, Vq = rd 
+
+  # mass matrix, tuple of differentiation matrices
+  @unpack M, Drst = rd
+
+  # ∫f(u) * dv/dx_i = ∑_j (Vq*D_i)'*diagm(wq)*(rstxyzJ[i,j].*f(Vq*u))
+  invMQrstTrW = map(D -> -M\((Vq*D)'*diagm(wq)), Drst)
```

Comment:
It's the inverse mass matrix (`invM`) multiplied against the transpose of `Qr`, `Qs`, `Qt` (essentially weak-form differentiation matrices with respect to the `r,s,t` coordinates) multiplied also by the diagonal matrix containing quadrature weights (`W`). 

I don't know how it could be any clearer `¯\_(ツ)_/¯` /s!

Reply:
I've renamed this to `weak_differentiation_matrices`. There's also `Qrst_skew_Tr` in `create_cache`; I'll rename that in the flux differencing volume kernel PR.